providers/base/bin/smartcard_test.py

```diff
@@ -0,0 +1,329 @@
+                return name
+        elif lt == "contactless":
+            if "contactless" in ln or "-cl" in ln:
+                return name
+        else:
+            return name
+
+    def list_readers(self, list_type: str):
+        """
+        List smart card readers in stringified format for the resource job
+
+        :param list_type: contact/contactless/ALL
+
+        """
+        for r in self.readers:
+            if self.reader_filter(list_type, r):
+                print(
+                    "smartcard_reader: {}".format(
+                        self.stringify_reader_name(r.name)
+                    )
+                )
+
+    def detect_reader(self, list_type: str):
+        """
+        Detect smart card readers
+
+        :param list_type: contact/contactless/ALL
+
+        """
+        count = 0
+        for r in self.readers:
+            if self.reader_filter(list_type, r):
+                self.logger.info(r.name)
+                count = count + 1
+        if count < 1:
+            raise SystemExit("There is no smartcard reader in this system")
+
+    def get_real_reader_instance(self, reader: str):
+        """
+        Using the stringified smartcard reader name
+        to get the actual reader instance
+
+        :param reader: Stringified smart card reader name
+        """
+        for r in self.readers:
+            if self.stringify_reader_name(r.name) == reader:
+                return r
+
+    def get_connection(self, reader: str):
+        """
+        Connect to the smartcard reader
+
+        :param reader: Stringified smart card reader name
+        """
+        sc_reader = self.get_real_reader_instance(reader)
+        if sc_reader:
+            try:
+                connection = sc_reader.createConnection()
+                connection.connect()
+                self.logger.info("[{}] connected".format(sc_reader))
+                return connection
+            except (NoCardException, CardConnectionException):
+                raise SystemExit("no card inserted or card is unsupported")
+        raise SystemExit("no smartcard reader")
+
+    @timeout(30)
+    def detect_smartcard(self, reader: str):
+        """
+        Detect smartcard insertion and removal in the smartcard reader
+
+        :param reader: Stringified smart card reader name
+        """
+        real_reader = self.get_real_reader_instance(reader)
+
+        self.logger.info(
+            "Smartcard insertion and removal detection test is starting"
+        )
+        self.logger.info(
+            "Please insert and remove the smartcard within 30 seconds.\n"
+        )
+
+        cardrequest = CardRequest(timeout=30, newcardonly=True)
+        cards = []
+        while len(cards) == 0:
+            currentcards = cardrequest.waitforcardevent()
+            for card in currentcards:
+                if (
+                    not cards.__contains__(card)
+                    and real_reader.name == card.reader
```

Comment:
This comparison might need an additional `isinstance` call since `card.reader` is `str | smartcard.reader.Reader.Reader` according to the [source](https://github.com/LudovicRousseau/pyscard/blob/master/src/smartcard/Card.py#L65-L67)

Reply:
also 

```py
if card not in cards
```